src/main/java/core/basesyntax/service/impl/DataConverterImpl.java

```
@@ -0,0 +1,83 @@
+        if (lines == null || lines.isEmpty()) {
+            throw new RuntimeException("Input parameter cannot be null or empty");
+        }
+        //header check
+        String header = lines.get(0);
+        if (!header.equals(HEADER_FORMAT)) {
+            throw new RuntimeException("Invalid file header. "
+                    + "Expected 'type,fruit,quantity' but found: " + lines.get(0));
+        }
+
+        List<FruitTransaction> fruitTransactions = new ArrayList<>();
+        for (int i = 1; i < lines.size(); i++) {
+            String currentLine = lines.get(i).trim();
+            if (currentLine.isEmpty()) {
+                continue;
+            }
+            fruitTransactions.add(parseLineToTransaction(currentLine));
+        }
+        return fruitTransactions;
+    }
+
+    private FruitTransaction parseLineToTransaction(String line) {
+
+        String[] parts = line.split(",");
+        //type/fruit/quantity validation
+        if (parts.length != 3) {
+            throw new RuntimeException("Invalid line format. Expected 3 columns but found: "
+                    + parts.length);
+        }
+
+        ////get correct operation using helper method
+        FruitTransaction.Operation operation = getOperationByCode(parts[0]);
+        //get correct fruitName using helper method
+        String fruitName = parseFruitName(parts[1]);
+        //get correct quantity using helper method
+        int quantity = parseQuantity(parts[2]);
+
+        return new FruitTransaction(operation, fruitName, quantity);
+    }
+
+    private FruitTransaction.Operation getOperationByCode(String operationCode) {
+        String code = operationCode.trim();
+        for (FruitTransaction.Operation operation : FruitTransaction.Operation.values()) {
+            if (operation.getCode().equals(operationCode)) {
+                return operation;
+            }
+        }
+        throw new RuntimeException("Input line is invalid: unknown operation code");
```

Error messages throughout parsing are somewhat generic (e.g., "Input line is invalid: unknown operation code"). The checklist and task description require informative RuntimeExceptions that include context such as file/line number and invalid value. Update all thrown RuntimeExceptions to include such context. For example: "Unknown operation code 'x' at input line 5: 'x,apple,10'". See checklist guidance about not ignoring exceptions and rethrowing informative RuntimeExceptions.